compiler/parser/RdrHsSyn.hs

```diff
@@ -2569,24 +2581,30 @@ mkLambda args body mBinds =
 
 
 mkChoiceDecls :: SrcSpan -> Located RdrName -> LHsLocalBinds GhcPs -> CombinedChoiceData -> [LHsDecl GhcPs]
-mkChoiceDecls templateLoc conName binds (CombinedChoiceData controllers ChoiceData{..} choiceTyVars flexible) =
-    [ noLoc (SigD noExt (TypeSig noExt [noLoc name] (mkHsWildCardBndrs (mkHsImplicitBndrs $ noLoc $ HsTupleTy noExt HsBoxedOrConstraintTuple [controllerSig, actionSig, consumingSig]))))
-    , noLoc (ValD noExt (FunBind noExt (noLoc name) (matchGroup noSrcSpan $ matchWithBinds (matchContext $ noLoc name) [] noSrcSpan (noLoc $ ExplicitTuple noExt (map (noLoc . Present noExt) [controllerDef, actionDef, consumingDef]) Boxed) (noLoc emptyLocalBinds)) WpHole []))
+mkChoiceDecls templateLoc conName binds (CombinedChoiceData controllers observersM ChoiceData{..} choiceTyVars flexible) =
+    [ noLoc (SigD noExt (TypeSig noExt [noLoc name] (mkHsWildCardBndrs (mkHsImplicitBndrs $ noLoc $ HsTupleTy noExt HsBoxedOrConstraintTuple [controllerSig, actionSig, consumingSig, observersSig]))))
+    , noLoc (ValD noExt (FunBind noExt (noLoc name) (matchGroup noSrcSpan $ matchWithBinds (matchContext $ noLoc name) [] noSrcSpan (noLoc $ ExplicitTuple noExt (map (noLoc . Present noExt) [controllerDef, actionDef, consumingDef, observersDef]) Boxed) (noLoc emptyLocalBinds)) WpHole []))
     ]
     where
         name = mkRdrUnqual $ mkVarOcc ("_choice_" ++ rdrNameToString conName ++ rdrNameToString cdChoiceName)
         consumingSig = (unLoc . mkQualType . show . fromMaybe Consuming <$> cdChoiceConsuming) `mkAppTy` templateType
         consumingDef = unLoc . mkQualVar . mkDataOcc . show . fromMaybe Consuming <$> cdChoiceConsuming
         controllerSig = mkFunTy templateType (mkFunTy choiceType partiesType)
         controllerDef = mkLambda controllerDefArgs controllers (Just (extendLetBindings binds (dummyBinds controllerDefArgs)))
-        controllerDefArgs = [this, controllerArg]
+        controllerDefArgs = [this, flexibleArg]
+        observersSig = mkMaybe (mkFunTy templateType (mkFunTy choiceType partiesType))
+        observersDef = case observersM of
+          Nothing -> eNothing
+          Just observers ->
+            eJust $ mkLambda observerDefArgs observers (Just (extendLetBindings binds (dummyBinds observerDefArgs)))
+        observerDefArgs = [this, flexibleArg]
         actionSig = mkFunTy contractIdType (mkFunTy templateType (mkFunTy choiceType choiceReturnType))
         actionDef = mkLambda actionDefArgs cdChoiceBody (Just (extendLetBindings binds (dummyBinds actionDefArgs)))
         actionDefArgs = [self, this, arg]
         arg = argPatOfChoice (noLoc $ choiceNameToRdrName $ mkDataOcc choiceName) cdChoiceFields
         choiceName = rdrNameToString cdChoiceName
         choiceNameToRdrName = if choiceName == "Archive" then qualifyDesugar else mkRdrUnqual
-        controllerArg = if flexible then arg else WildPat noExt
+        flexibleArg = if flexible then arg else WildPat noExt
```

Comment:
If we change this, let's please just inline it. Coming up with a _good_ name for this is too hard for such a small code snippet to justify any mental effort.

Some probably pointless hair-splitting: In my understanding, which seems to be matched by [Wikipedia](https://en.wikipedia.org/wiki/Parameter_(computer_programming)#Parameters_and_arguments), this is not an _argument_ but a _(formal) parameter_ (which some people call _formal argument_ but then the _formal_ is important).

Reply:
`flexibleArg` is now used in two places, so inlining it will cause duplication, which is not an improvement in my opinion.

I assume you don't actually want me to rename all the things called `Arg` to `Parameter`.

Reply:
Oh, I missed the second occurrence. In fact, it shouldn't exist in the first place. Let's remove fix the other occurrence as indicated above and inline `flexibleArg` then.

Reply:
done